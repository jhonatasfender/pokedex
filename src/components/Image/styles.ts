import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import styled from "styled-components";

export const Icon = styled(FontAwesomeIcon)`
  position: absolute;
  top: 4rem;
  right: 4rem;
  width: 3rem;
  height: 3rem;
  opacity: 0.5;
`;
