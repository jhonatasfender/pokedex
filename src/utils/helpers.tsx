import React from "react";
import { ThemeProvider } from "styled-components";

import { render, RenderResult } from "@testing-library/react";

import theme from "../style-global/theme";

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
