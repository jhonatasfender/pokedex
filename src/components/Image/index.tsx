import { HTMLAttributes, useState } from "react";

import { faQuestion } from "@fortawesome/free-solid-svg-icons";

import { Sprites } from "../../hooks/use-pokemon";
import * as S from "./styles";

type ImageProps = {
  sprites: Sprites;
} & Omit<HTMLAttributes<HTMLImageElement>, "onLoad">;

export default function Image({ sprites, ...props }: ImageProps) {
  const [usingDefault, setDefault] = useState(false);

  const imageOnError = () => {
    setDefault(true);
  };

  return !usingDefault ? (
    <img {...props} src={sprites.normal} alt="pokemon" onError={imageOnError} />
  ) : (
    <S.Icon icon={faQuestion} />
  );
}
