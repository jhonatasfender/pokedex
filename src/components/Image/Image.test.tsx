import { fireEvent, screen, waitFor } from "@testing-library/react";
import Image from ".";

import { renderWithTheme } from "../../utils/helpers";

describe("Header", () => {
  it("should initialize with default parameters", () => {
    renderWithTheme(
      <Image
        sprites={{
          animated:
            "https://img.pokemondb.net/sprites/black-white/anim/normal/bulbasaur.gif",
          large: "https://img.pokemondb.net/artwork/bulbasaur.jpg",
          normal:
            "https://img.pokemondb.net/sprites/omega-ruby-alpha-sapphire/dex/normal/bulbasaur.png",
        }}
      />
    );

    const img = screen.getByRole("img");

    expect(img).toMatchSnapshot();
  });

  it("should initialize with default parameters with error", async () => {
    renderWithTheme(
      <Image
        sprites={{
          animated:
            "https://img.pokemondb.net/sprites/black-white/anim/normal/bulbasaur.gif",
          large: "https://img.pokemondb.net/artwork/bulbasaur.jpg",
          normal:
            "https://img.pokemondb.net/sprites/omega-ruby-alpha-sapphire/dex/normal/bulbasaur.pn",
        }}
      />
    );

    const img = screen.getByRole("img");

    fireEvent.error(img);

    await waitFor(() => {
      expect(screen.queryByRole("img")).not.toBeInTheDocument();
    });

    expect(img).toMatchSnapshot();
  });
});
