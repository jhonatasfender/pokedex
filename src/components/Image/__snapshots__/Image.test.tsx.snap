// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should initialize with default parameters 1`] = `
<img
  alt="pokemon"
  src="https://img.pokemondb.net/sprites/omega-ruby-alpha-sapphire/dex/normal/bulbasaur.png"
/>
`;

exports[`Header should initialize with default parameters with error 1`] = `
<img
  alt="pokemon"
  src="https://img.pokemondb.net/sprites/omega-ruby-alpha-sapphire/dex/normal/bulbasaur.pn"
/>
`;
